### ISSUE TITLE
`chat_template` Strategy Never Trains the Gemma Turn Terminator

### ISSUE THREAD

**Issue by thad0ctor** [axolotl-ai-cloud/axolotl#3754]
Credit: @ferno9 for discovering the issue and diagnosing 

### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

When fine-tuning any **Gemma** model (1 / 2 / 3 / 3n / 4) with `type: chat_template`, the per-turn stop token that the Gemma chat template actually emits — `<end_of_turn>` (id **107** on Gemma 1/2, id **106** on Gemma 3/3n) or `<turn|>` (id **106** on Gemma 4) — should be **trained** at the end of each assistant turn (it should receive a real label, honoring `train_on_eos`/`train_on_eot`), so the model learns to stop. This already happens by default for other families (Qwen `<|im_end|>`, Llama-3 `<|eot_id|>`, Mistral `{{ eos_token }}`, …).

### Current behaviour

For Gemma, the turn terminator is **silently masked (`-100`) by default**, so the model is never taught to stop. After fine-tuning, the model produces the correct response. Loss/eval curves look perfectly healthy because the only thing never learned — stopping — is never in the loss.

**Root cause:** `ChatTemplateStrategy` defaults `eot_tokens` to `[tokenizer.eos_token]`, and the EOS/EOT finders key off `tokenizer.eos_token_id`. For Gemma that is `<eos>` (id **1**), but `<eos>` **never appears in the rendered chat** (turns end with `<end_of_turn>`/`<turn|>`), so neither finder matches the turn boundary → the terminator's label stays `-100`.

A warning *is* emitted but is easy to miss and not an error:

```
[WARNING] [axolotl.prompt_strategies.chat_template] EOS token '<eos>' not found in chat_template. Please check if your template/EOS token is correct.
```
The default gemma configs (e.g. Gemma 4) don't include the config flags needed to train the real stop token (the most likely fine-tuning scenario for non-base variants). The override works however:

```
  special_tokens:
    eos_token: "<turn|>"     # "<end_of_turn>" for gemma 2/3/3n
  train_on_eos: turn

```

**Why other models are fine:** their `tokenizer.eos_token` *is* the turn terminator (Qwen → `<|im_end|>`, Llama-3-instruct → `<|eot_id|>`) or the template renders the `{{ eos_token }}` variable. Gemma is the only family where HF keeps `tokenizer.eos_token = <eos>` and exposes the real terminator only via `generation_config.eos_token_id` (e.g. gemma-4-31B-it = `[1, 106, 50]`, gemma-3-4b-it = `[1, 106]`). So the eos-keyed default misses it for **every Gemma generation**.

Relevant code: `src/axolotl/prompt_strategies/chat_template.py`
- `eot_tokens` defaults to `[self.tokenizer.eos_token]` (~L310–317)
- `_validate_eot_and_eos_tokens` warns but does not fix (~L350–374)
- label assignment via `find_first_eos_token` / `find_first_eot_token`, both keyed on token id 1 (~L599–637)

## Steps to reproduce

1. Create a tiny chat dataset `train.jsonl` (2 lines is enough), e.g.:
   ```json
   {"messages":[{"role":"user","content":"Translate to French: The weather is nice today."},{"role":"assistant","content":"Il fait beau aujourd'hui."}]}
   {"messages":[{"role":"user","content":"Translate to French: We met at the library."},{"role":"assistant","content":"Nous nous sommes rencontrés à la bibliothèque."}]}
   ```
2. Use the minimal config below (the official `examples/gemma4/31b-qlora.yaml` reproduces it identically — it sets neither `eot_tokens` nor `special_tokens.eos_token`).
3. Run preprocess with label debug:
   ```bash
   python -m axolotl.cli.preprocess config.yaml --debug 2>&1 | grep -E "not found in chat_template|turn\|>\("
   ```
4. **Observe** the warning and the masked terminator on every assistant turn:
   ```
   ... aujourd'hui .(236761, 236761) <turn|>(-100, 106)     <-- masked, never trained
   ```
   (Same result with `ds_type: json` or `ds_type: parquet`, and with `sample_packing: true` or `false` — each packed sample's `<turn|>` is masked too, so packing is not involved.)

5. **Confirm the workaround** trains it — add to the config:
   ```yaml
   special_tokens:
     eos_token: "<turn|>"   # use "<end_of_turn>" for Gemma 1/2/3/3n
   train_on_eos: turn
   ```
   Re-run step 3 → warning is gone and the terminator is now trained:
   ```
   ... la bibliothèque .(236761, 236761) <turn|>(106, 106)   <-- trained
   ```

### Config yaml

```yaml
base_model: google/gemma-4-31B-it
chat_template: gemma4
datasets:
  - path: ./train.jsonl
    ds_type: json
    type: chat_template
    field_messages: messages
adapter: lora
lora_r: 16
lora_alpha: 16
lora_target_modules: 'model.language_model.layers.[\d]+.(_checkpoint_wrapped_module.)?(mlp|self_attn).(up|down|gate|q|k|v|o)_proj'
sequence_len: 2048
micro_batch_size: 1
num_epochs: 2
# No special_tokens / eot_tokens set -> reproduces the bug (matches examples/gemma4/31b-qlora.yaml)
```

### Possible solution

If the intended behavior is to reduce code complexity the example configs could be updated to note the overrides are required for training on EOS otherwise.

More robust fix (mirrors how Unsloth and LLaMA-Factory handle it):

1. **Derive the trainable terminator from the rendered chat template** instead of `tokenizer.eos_token`. Render the active template on a dummy assistant turn and read the token(s) emitted **after** the assistant content; default `eot_tokens` to that token when the user hasn't set it. The tokenizer resolves the correct id automatically, so this needs **no hardcoded ids/strings** and handles `<end_of_turn>`=107 (Gemma 1/2) vs 106 (Gemma 3/3n) vs `<turn|>` (Gemma 4) transparently. It also generalizes to any other model with an `eos_token` ≠ terminator split.
   - Optional secondary cross-check: `generation_config.eos_token_id` (lists the terminator on Gemma 3/3n/4). Do **not** rely on it alone — Gemma 1 and gemma-2-9b/27b-it omit it, and Gemma 3 *base* also lists `[1,106]`, so it's not a reliable base-vs-instruct signal.

2. **Gate on chat-template presence** This is essentially free (the `chat_template` strategy only runs when a template is in play) and avoids mis-handling non-chat Gemma derivatives. 

3. **Keep the existing override** (`eot_tokens` / `special_tokens.eos_token`) as the always-wins escape hatch — just make the default correct so it isn't required.

4. **(Optional, recommended) also override `tokenizer.eos_token` → terminator** (cf. LLaMA-Factory `replace_eos`, Unsloth `map_eos_token`) so the **saved** adapter/model stops at inference too. The train-side and export-side fixes are independent; doing both prevents "answer then never stops" at train *and* serve time.

A minimal regression test: preprocess a Gemma chat sample and assert the terminator token id receives a non-`-100` label (and the `<eos> not found` warning is gone).

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12

### axolotl branch-commit

main/4d6490b

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by thad0ctor** [comment]
Happy to tackle this if you guys would like - the more robust fix would avoid the need to remember to overlook future overrides if newer gemma models carry the same behavior

**Comment by NanoCode012** [comment]
I think: we should update the example configs to set `eot_token: <|turn|>` or whichever respective token at least first? This config was built for this. Overriding the eos_token also works though.

> Derive the trainable terminator from the rendered chat template instead of tokenizer.eos_token

Trickier. Some templates do not have a EOT per se. Some just outputs the next turn's tag (for ex: `<human>`) and uses it as the stopping point. I think mertharme (which I understand is old) falls into this. Templates like chatml `im_end` is easier.

Maybe the better approach would be to make sure we check this on future model support. I used to do this, but missed it for recent models..

> Gate on chat-template presence 
What would we even gate? We can't tell if a template's "Gemma". All we get is the jinja string.

> (Optional, recommended) also override tokenizer.eos_token → terminator 

I'd not want the eot_token to override the eos_token. Usually these inference engine should support multiple stop tokens either way? I'm more interested if we can inject this into configs to let them be picked up downstream. (I believe generation_config now supports eos_token: List, unless I remembered incorrectly)

**Comment by NanoCode012** [comment]
I believe this was fixed, correct?

**Comment by thad0ctor** [comment]
> I believe this was fixed, correct?

@NanoCode012 yes, the immediate term fix was https://github.com/axolotl-ai-cloud/axolotl/pull/3756

https://github.com/axolotl-ai-cloud/axolotl/pull/3762 attempted to root cause it with chat templates but it is potentially regression-prone
